### DISCUSSION TITLE
e107 session table

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#4992] [Q&A Core] [answered]
Hi,
I noticed a lot of records on site that is almost not used.  Table e107_session has over 3500 records with valid 24 hours sessions.  

So I checked my other active site. The specific of this site is that 2 CMS are running for the same user set.  One is very similar as old e107.  

Records in old table looks correctly, new e107 session table looks wrong.  

![image](https://user-images.githubusercontent.com/5429548/232195588-7f382398-1ab0-4c87-96dc-be59779cfbe8.png)  

Could you explain me this, please?  Thanks

**Comment by Deltik** [comment, marked as answer]
e107 currently saves the sessions of guest visitors.  Bots tend not to persist cookies, so every time a bot hits your site, a row gets added.  We could definitely improve this by not saving cookies for guests.

Unfortunately, one usage I can think of that would break from not saving cookies is `/e107_images/secimg.php`, which [saves the answer to the CAPTCHA inside the session](https://github.com/e107inc/e107/blob/c3517641e07d4bd6427a79ee0530b7173c1c4c94/e107_handlers/secure_img_handler.php#L126).  A better approach for this would be to send the client a cryptographically signed string that only the server can decode so that the CAPTCHA answer would not depend on cookies and sessions.

**Reply by Jimmi08** [reply]
Thank you for your answer. 

What I found:

- that other software is saving data for bots too. I tested with the google search console and yes, the record for Googlebot IP was created too. 

- it deletes outdated sessions asap - this didn't work with e107 for me, I tried to set the lifetime to 100 or something and records were still there.  I couldn't find this in the session handler (the code is there, but when it is fired ).  Edit: is possible that this works only for admins or in the admin area or something like that?   

The problem was fixed after adding in class2.php:    (the user test was added after the problem with unauthorized access)
`    e107::getDb("session")->delete("session", '`session_user` = 0 AND `session_expires`<'. $past ); `

As soon as I added this, issue with unathorized access occured in saving prefs (I wasn't able to set 100 back to 3600, I had to do this via phpmyadmin).  The problem was caused by session_regenerate_id() in shutdown() method.  

But I think that the main difference is that e107 saves records by session_id and UN by IP or username.  Each tab creates a new session ID  for the same user...    

I truncated both tables when starting this comment - actually UN table 75 records, e107 140 records. UN dashboard is saying 75 users (19 members + 56 quests). e107 dashboard is saying 18 users (5 members + 7 bots + rest).  

When I did this yesterday, with lifetime 3600s , it was cca 200 records vs 3500 after 12 hours.

**Reply by Deltik** [reply]
By the way, `e107_session.session_user` is not hooked up to anything right now due to an implementation bug.  It would have been if https://github.com/e107inc/e107/pull/4176/files#diff-266827643205142c75b79ab6f4dae6b313952873f6e3ed11f08420535c97ca5bR187 were merged, but my merge request was not approved, and an incomplete implementation (https://github.com/e107inc/e107/commit/7439d599f84d122df06dae61ea09ce3e0415d8af) was pushed instead.

**Reply by Jimmi08** [reply]
Hm, it is dissapointment.  I planned to replace old stuff with e107 version, not only here, but with other cms too. 
I suppose e107 changes are related to safety so I wanted to do the same changes. Now it has to wait. 

By the way.  After changing lifetime to 3600 (to have it the same), troubles with admin started (no files changes).  I was logged in, I left admin dashboard opened,  and after some hours there was white page. I was still logged in, just admin.php was blank. So I returned it back to 86400 (24hours). I had to log out and then log in back.   

I think it should delete all quest records after 3600 (by default, ignore lifetime) and accept lifetime for user logged in (it should be replacement for remember me).   

Thanks for your time

**Reply by Jimmi08** [reply]
@Deltik Can you point me in your version where db records for guests are deleted?  I found it only in logout process. 
Thanks

**Reply by Deltik** [reply]
There is no special cleanup for unauthenticated visitors.  The built-in PHP session garbage collection ([`SessionHandler::gc()`](https://www.php.net/manual/en/sessionhandler.gc.php)) is used for all sessions.

**Reply by Jimmi08** [reply]
yes, and it is correct,  Just, it should be called with close() but I couldn't find where this (close()|  is called in e107.

**Reply by Deltik** [reply]
[`SessionHandler::close()`](https://www.php.net/manual/en/sessionhandler.close.php) is automatically called when the session is closed.  e107 kinda does this with [`e_session::end()`](https://github.com/e107inc/e107/blob/c3517641e07d4bd6427a79ee0530b7173c1c4c94/e107_handlers/session_handler.php#L877-L886), but it happens quite late in [`e107_core/templates/footer_default.php`](https://github.com/e107inc/e107/blob/c3517641e07d4bd6427a79ee0530b7173c1c4c94/e107_core/templates/footer_default.php#L468-L468) due to https://github.com/e107inc/e107/issues/1446, which is a performance-harming workaround for the CAPTCHA depending on sessions (as I mentioned previously).

The current workaround for the performance loss of moving session closing to the end is not doing any locking per session when using `e_session_db`.  This should be "okay" for the vast majority of visits, as we don't store that much stuff in sessions that could conflict from multiple visits.

**Reply by Jimmi08** [reply]
Thanks, finally I got it. This is used:  https://www.php.net/manual/en/class.sessionhandlerinterface.php   With my PHP limitation I couldn't get how it could work :)

**Comment by Jimmi08** [comment]
The code example from gitter discussion:

```
<?php
require_once('e107_config.php');

$gc_maxlifetime = (int)ini_get('session.gc_maxlifetime');
if (!$gc_maxlifetime) {
    $gc_maxlifetime = 3600;
}

$expiration_time = time() - $gc_maxlifetime + 3600; // One hour

$dsn = "mysql:host=$mySQLserver;dbname=$mySQLdefaultdb;charset=$mySQLcharset";
$pdo = new PDO($dsn, $mySQLuser, $mySQLpassword);

$table_name = $mySQLprefix . "session";

$query = "DELETE FROM `$table_name` WHERE (session_user = 0 OR session_user IS NULL) AND session_expires < $expiration_time LIMIT 1000";

$total_deleted = 0;
do {
    $stmt = $pdo->prepare($query);
    $stmt->execute();
    $deleted_rows = $stmt->rowCount();
    $total_deleted += $deleted_rows;
} while ($deleted_rows > 0);

echo "Deleted " . $total_deleted . " rows from the $table_name table.\n";
```
